stdlib/LinearAlgebra/src/diagonal.jl

```diff
@@ -10,9 +10,20 @@ struct Diagonal{T,V<:AbstractVector{T}} <: AbstractMatrix{T}
         new{T,V}(diag)
     end
 end
+Diagonal{T,V}(d::Diagonal) where {T,V<:AbstractVector{T}} = Diagonal{T,V}(d.diag)
 Diagonal(v::AbstractVector{T}) where {T} = Diagonal{T,typeof(v)}(v)
 Diagonal{T}(v::AbstractVector) where {T} = Diagonal(convert(AbstractVector{T}, v)::AbstractVector{T})
 
+function Base.promote_rule(A::Type{<:Diagonal{<:Any,V}}, B::Type{<:Diagonal{<:Any,W}}) where {V,W}
+    X = promote_type(V, W)
+    # We check if the promoted type has a well-defined eltype.
+    # if it does, use it to construct the Diagonal type
+    _promoted_type(::Type{T}, ::Type{Z}) where {T,Z<:AbstractVector{T}} = Diagonal{T,Z}
+    # If it doesn't, we fall back to the common ancestor
+    _promoted_type(::Type, ::Type) = typejoin(A, B)
```

Comment:
This is a hard-to-understand way of implementing this; it would be clearer to do this via runtime logic. Since `promote_rule` gets run inside a `@pure` caller, I don't think there's any negative to that.